Adding categorical embeddings

lastmansleeping opened the pull request.

lastmansleeping commented on python/ml4ir/base/model/scoring/interaction_model.py:
```diff
@@ -51,21 +49,19 @@ def __init__(
 
     def transform_features_op(self, train_features, metadata_features):
-        # TODO: Make train_features a dictionary
+        # Sorting so that we control the order
```
This allows users to do more complex customization if needed.

Ullimague replied:
We need to add a comment as such, and why we have created transform_features_op.

lastmansleeping commented on python/ml4ir/base/features/feature_layer.py:
```diff
@@ -1,14 +1,12 @@
-# flake8: noqa
```
Fixed code complexity and consolidated the feature layer methods into one single method that handles both SequenceExamples and Examples

lastmansleeping commented:
Added 3 different kinds of categorical embeddings:
- using hash buckets for string features
- using vocabulary files
- using numeric indices

Also cleaned up the feature layer and consolidated many methods.

TODO:
- adding unit testing
- verify that vocabulary is persisted in SavedModel and add test for same

FYI @jakemannix

Ullimague commented on python/ml4ir/applications/ranking/tests/data/config/feature_config.yaml:
```diff
@@ -106,15 +106,46 @@ features:
+      required: true
       default_value: 0
+    tfrecord_type: context
+  - name: group_name
```
What's the difference between group and group_name?

Ullimague commented on python/ml4ir/applications/ranking/tests/test_feature_processing.py:
```diff
@@ -0,0 +1,31 @@
+        )
+
+        # Overall
+        assert processed_text.replace("\x00", "") == input_text.lower().translate(
```
Why do we need to replace \x00 (null character?)

Ullimague commented on python/ml4ir/applications/ranking/tests/test_feature_processing.py:
```diff
@@ -0,0 +1,31 @@
+from ml4ir.base.features import preprocessing
+
+
+class RankingModelTest(RankingTestBase):
```
Why should this be in rankin module, when we test the base package?
Do you plan to have all the tests ins /ranking?

Ullimague commented on python/ml4ir/applications/ranking/tests/test_feature_processing.py:
```diff
@@ -0,0 +1,31 @@
+class RankingModelTest(RankingTestBase):
+    def test_text_preprocesing(self):
+        """
+        Unit test text preprocessing
```
Can we make explicit in comments that here we test:
- punctuation removal
- lower casing

Ullimague replied:
Don't forget this.

Ullimague commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,200 @@
+
+
+def categorical_embedding_with_hash_buckets(feature_tensor, feature_info):
+    """Embedding lookup for categorical features"""
```
We would need an extensive doc (same for all methods in this file).

Ullimague commented on python/ml4ir/base/io/file_io.py:
```diff
@@ -45,6 +45,17 @@ def make_directory(dir_path: str, clear_dir: bool = False, log=None) -> str:
 
 
+def read_list(infile: str):
+    """Read a text file"""
```
It does more that just read, it converts it to a list + do manipulations. Could we be explicit in comments about the assumptions?

Ullimague commented on python/ml4ir/base/features/feature_fns/sequence.py:
```diff
@@ -0,0 +1,41 @@
+    feature_tensor = tf.reshape(feature_tensor, [-1, feature_layer_info["args"]["max_length"]])
+    if "embedding_size" in feature_layer_info["args"]:
+        char_embedding = layers.Embedding(
+            input_dim=256,
```
Do we plan to have input_dim parametrizable? Should we make it global variable as it appears in both conditions?

lastmansleeping commented:
@Ullimague @balikasg Updating the PR with the final changes:
- addressed the initial PR comments
- added unit tests for all of the feature functions
- added documentation to the tests and the feature functions as requested

Created a new Github issue to rearrange the tests so that we move most of the generic tests under `ml4ir/base`. They are currently under `ml4ir/applications/ranking`.

Ullimague commented on python/ml4ir/applications/ranking/tests/test_feature_layer.py:
```diff
@@ -0,0 +1,166 @@
+        """
+        Goal:
+        Convert a string tensor to its corresponding sequence encoding
+        obtained through the bytes_sequence_to_encoding fn
```
very minor suggestion: you can use "function (fn)" instead of "fn", + '.' at the end of the sentence.

Ullimague commented on python/ml4ir/applications/ranking/tests/test_feature_layer.py:
```diff
@@ -0,0 +1,166 @@
+class RankingModelTest(RankingTestBase):
+    def test_bytes_sequence_to_encoding(self):
+        """
+        Goal:
```
very minor suggestion: instead of "Goal:" we could "Asserts the conversion of a string tensor [...]"

lastmansleeping replied:
Will update.

Ullimague commented on python/ml4ir/applications/ranking/tests/test_feature_layer.py:
```diff
@@ -0,0 +1,166 @@
+    def test_categorical_embedding_with_hash_buckets(self):
+        """
+        Goal:
+        Convert a categorical string tensor into a categorical embedding
```
very minor suggestion: same as above.

Ullimague commented on python/ml4ir/base/features/feature_config.py:
```diff
@@ -348,6 +348,7 @@ def generate_mask(self):
         """Add mask information used to flag padded records"""
         return {
             "name": "mask",
+            "node_name": "mask",
```
Thanks @lastmansleeping!

Ullimague commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+    each of size hash_bucket_size, then converting each hash bucket into
+    a categorical embdding of dimension embedding_size. Finally, these embeddings
+    are combined either through mean, sum or concat operations to generate the final
+    embedding.
```
awesome, very small suggestion: [...] the final embedding + based on the feature_info + .

Ullimague commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+
+    Args:
+        feature_tensor: String feature tensor
+        feature_info: Dictionary representing the feature_info for the specific feature from the FeatureConfig
```
very small suggestion: "Dictionary representing the specific feature from the FeatureConfig." (removing "feature_info for the", we shouldn't define a word (feature_info) by itself :))

lastmansleeping replied:
Good catch.

Ullimague commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+    Returns:
+        categorical embedding for the input feature_tensor
+
+    Args under feature_layer_info:
```
Great!!

Ullimague commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+        elif feature_layer_info["args"]["merge_mode"] == "concat":
+            embedding = tf.concat(
+                embeddings_list,
+                axis=-1,
```
out of curiosity, why do we have a "axis=-1" for this one?

lastmansleeping replied:
Because it is concatenating on the last axis which has the embeddings, whereas the other operations are reduce operations on the first axis. It is a strange tensorflow usage.

```
In [20]: tf.concat([np.random.randn(128, 16)] * 4, axis=1).shape
Out[20]: TensorShape([128, 64])

In [21]: tf.reduce_sum([np.random.randn(128, 16)] * 4, axis=0).shape
Out[21]: TensorShape([128, 16])
```

Ullimague commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+
+    Args:
+        feature_tensor: int feature tensor
+        feature_info: Dictionary representing the feature_info for the specific feature from the FeatureConfig
```
very small suggestion: same as above.

Ullimague commented on python/ml4ir/base/features/feature_fns/sequence.py:
```diff
@@ -0,0 +1,65 @@
+        merge_mode="concat",
+    )(char_embedding)
+    encoding = tf.expand_dims(encoding, axis=1)
+    # if feature_info.get("tfrecord_type") == SequenceExampleTypeKey.CONTEXT:
```
Should we keep this?

lastmansleeping replied:
I have kept it for historical context. Can remove actually.

jakemannix replied:
yeah, we can always pull from git history if needed.

lastmansleeping replied:
Removed.

Ullimague commented on python/ml4ir/applications/ranking/tests/test_feature_layer.py:
```diff
@@ -0,0 +1,54 @@
+from ml4ir.base.config.keys import SequenceExampleTypeKey
+
+
+# class RankingModelTest(RankingTestBase):
```
Do you plan on uncommenting this? What's the status?

lastmansleeping replied:
Looks like this is "outdated" code. I have updated this module with the the python unit tests for the feature layer.

Ullimague commented on python/ml4ir/base/features/feature_layer.py:
```diff
@@ -19,8 +17,10 @@ class FeatureLayerMap:
 
```
We still need to fix this.

lastmansleeping replied:
I have fixed this. Looks like it is "outdated" as well. Not sure why this is happening.

lastmansleeping commented:
@Ullimague @balikasg 
You can test the code by running the following command:
```
python ml4ir/applications/ranking/pipeline.py \
--data_dir ml4ir/applications/ranking/tests/data/tfrecord \
--feature_config ml4ir/applications/ranking/tests/data/tfrecord/feature_config.yaml \
--run_id test \
--data_format tfrecord \
--execution_mode train_inference_evaluate
```

It does the following:
- load feature_config from path
- load TFRecord data as SequenceExample from `data_dir`
- define a Keras model using default model config
- train, validate and test the model on the loaded data respectively
- predict the model ranking scores and save to file
- save the model with both default and tfrecord serving signatures

balikasg commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+        raise NotImplementedError
+
+    # String input features
+    elif feature_info["dtype"] in (tf.string,):
```
The code exits `if feature_info["dtype"] not in (tf.float32, tf.int64)` e.g., `if not feature_info["dtype"] ==tf.string; raise ..`
Given the above, the `elif` statement can be removed and the code below visually simplified removing indentation.

lastmansleeping replied:
That's true. I should probably just remove the `tf.float32` and `tf.int64` from this method as well. Not really needed.

balikasg commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+            )
+
+        embedding = None
+        if feature_layer_info["args"]["merge_mode"] == "mean":
```
Since you are using if/elif (can do it with several ifs) you can use the `else` statement to capture anything different. Like the use uses a type `maen` and you raise an error message "We currently support mean/sum/concat as modes for reducing hashed embeddings to a single one. You selected XX, please update your config."

balikasg replied:
More user friendly and actionable I mean, it makes the link between code and config... Keras outputs such messages that I like (personal opinion).

lastmansleeping replied:
Yes, sir. Will do this. Good call.

balikasg commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+
+def categorical_embedding_with_indices(feature_tensor, feature_info):
+    """
+    Convert input integer tensor into categorical embedding.
```
These are great improvements in the documentation @lastmansleeping @Ullimague 
Minor: Should be `Converts` as I guest the object implied is `the function or the code`. Similarly for other functions

balikasg commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+
+class VocabLookup(layers.Layer):
+    """
+    Class defines a keras layer wrapper around a tf lookup table using the given vocabulary list.
```
minor: The class..

balikasg commented on python/ml4ir/base/features/feature_fns/categorical.py:
```diff
@@ -0,0 +1,260 @@
+def categorical_embedding_with_vocabulary_file(feature_tensor, feature_info):
+    """
+    Converts a string tensor into a categorical embedding representation.
+    Works by using a vocabulary file to convert the string tensor into categorical indices
```
Also in the class definition above: Do we convert a string to explicitly one categorical index or several categorical indices or both are possible?

lastmansleeping replied:
It is one categorical index. So a 1->1. Alternatively, what I can do is read a file with the key and the index(value) mapping. That way we can make it many to one. Should we do that?

lastmansleeping replied:
I tested the many to one vocabulary lookup. It works.

balikasg commented on python/ml4ir/base/features/feature_fns/sequence.py:
```diff
@@ -0,0 +1,57 @@
+    else:
+        char_embedding = tf.one_hot(feature_tensor, depth=256)
+
+    encoding = layers.Bidirectional(
```
Are we explicitly assuming here that the user wants a character-level BiLSTM? If so, would it make sense to call the fn CharacterLevelBiLSTMEncoder (the name smells like a class, I get it)? With high probability I am missing context here, so feel free to disregard...

lastmansleeping replied:
No. You are right. Initially, I had the `bilstm` part as an argument so the user could control how you encode the sequence through the feature config, but then removed it because it was probably too generic. I think calling this a bilstm encoder is a great idea. I will do that.

balikasg commented on python/ml4ir/base/io/file_io.py:
```diff
@@ -45,6 +45,17 @@ def make_directory(dir_path: str, clear_dir: bool = False, log=None) -> str:
+    if infile.startswith("hdfs"):
+        raise NotImplementedError
+    else:
+        txt = open(infile, "r").read()
```
same as in another part of the code. You don't need the else. Can be simplified.

lastmansleeping replied:
Unfortunately, here I do need it. The plan is to add the `if` section soon(as I work on enabling deepsea support for ml4ir). I didn't want to implement it as part of the other stories is all.

balikasg commented on python/ml4ir/base/io/file_io.py:
```diff
@@ -45,6 +45,17 @@ def make_directory(dir_path: str, clear_dir: bool = False, log=None) -> str:
+    else:
+        txt = open(infile, "r").read()
+
+    return [token.strip() for token in txt.splitlines()]
```
the naming assumes you are given a file with a token per line here. 
There is an alternative fancy way, for text files. You can use `pathlib` and have `infile` to be a Path. [If so](https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_text): 
`infile.read_text().split("\n")` 
The advantage if that it takes care of closing the file, which is missing here.

lastmansleeping replied:
Did not know about this. I like it. Will do. Thanks!
